packages/format-library/src/link/inline.js

```diff
@@ -17,20 +18,27 @@ import {
 	getTextContent,
 	slice,
 } from '@wordpress/rich-text';
-import { URLPopover } from '@wordpress/block-editor';
+import { __experimentalLinkControl as LinkControl } from '@wordpress/block-editor';
 
 /**
  * Internal dependencies
  */
 import { createLinkFormat, isValidHref } from './utils';
 
-const stopKeyPropagation = ( event ) => event.stopPropagation();
+function InlineLinkUI( { isActive, activeAttributes, addingLink, value, onChange, speak, stopAddingLink } ) {
+	const [ mountingKey, setMountingKey ] = useState( uniqueId() );
 
-function isShowingInput( props, state ) {
-	return props.addingLink || state.editLink;
-}
+	useEffect( () => {
+		return () => stopAddingLink();
+	}, [] );
+
+	useEffect( () => {
+		// Should remount the popover to focus it properly when isAdddingLink changes
+		if ( addingLink ) {
+			setMountingKey( uniqueId() );
+		}
+	}, [ addingLink ] );
```

Comment:
I noticed in e2e tests that there's an expectation that the focus move back to the Popover if we hit "ctrl+k" on an existing link. This is a hack that does that (a similar more implicit hack was used before)

Reply:
Sounds good. We can iterate later on a better solution.

Reply:
> I noticed in e2e tests that there's an expectation that the focus move back to the Popover if we hit "ctrl+k" on an existing link. This is a hack that does that (a similar more implicit hack was used before)

There might be another option here, based on the fact that the issue occurs because the popover is shown both when the caret is within a link (`isActive`) and when adding/editing a link (`addingLink`). We want to shift focus into the popover when editing the link, but not when we simply select it (click on it, or use arrow keys to change selection into it).

So ultimately it seems a question of: When the Popover's `focusOnMount` prop [changes from `false` to `'firstElement'`](https://github.com/WordPress/gutenberg/blob/2d1add52ec4f179b5e2fca0814a48c255f84da2b/packages/format-library/src/link/inline.js#L112), should we trigger the focus if the popover is already mounted? That would be enough to resolve the issue, and could be a valid way to control the focus behavior while the popover is already mounted (that's what's happening here). The main concern is... well, the name. It's called `focusOnMount`, so it would be odd if we're enhancing it to support functionality specifically serving the use-case that the popover is already mounted 😅

Reply:
A compromise solution might be to manually shift focus back into one of the popover elements when starting to edit the link. I could see this being something where:

- Either `Popover` or `LinkControl` offer a `focus` method. React's [documented `useImperativeHandle` example](https://reactjs.org/docs/hooks-reference.html#useimperativehandle) is exactly what we'd want in this case.
- Or, the `InlineLinkUI` renders its own `ref`-referenced DOM wrapper to `LinkControl`, then tries to [find a focusable target](https://github.com/WordPress/gutenberg/blob/master/packages/dom/src/focusable.js) within the container to shift focus.

Reply:
Reading this, It doesn't seem like there's a perfect solution. I might think a `focus` method is maybe the most explicit and understandable solution. What do you think?

Reply:
I'm thinking the simplest "non-decision" option would just be to add a `div` wrapper within the popover content, wrapping `LinkControl`, then call `wp.dom.focus.focusable.find`. Seems like that would be pretty straight forward.

The main downsides might be:

- It may be a common pattern for this sort of use of LinkControl in Popover
   - If it becomes a worry, we could consider again some abstraction like URLPopover
- It duplicates a behavior inherent to Popover

Neither are really overly concerning to me. Maybe it could make sense to expose a `focus` from Popover at some point, if it's a common need. It could even be reused internally for the component's own `focusOnMount` behavior. I don't think we need to commit to that API now though. Similarly, creating a `ref` behavior with `LinkControl` isn't something I'd want to do without considering how to make this consistent with other `*Control` components.